### ISSUE TITLE
what is AutoModelForCausalLMWithValueHead?

### ISSUE THREAD

**Issue by akk-123** [lvwerra/trl#180]
trl use `AutoModelForCausalLMWithValueHead`，which is base_model(eg: GPT2LMHeadModel) + fc layer，but I can't understand why need a fc head layer?

**Comment by lvwerra** [comment]
In PPO optimization a value estimation for each action (=generated token) is required. The value head (which is just an additional fc layer) makes that estimation.

**Comment by akk-123** [comment]
ok, thanks

**Comment by xymtxwd** [comment]
Hi @lvwerra , thanks for the implementations! A n00b question: in my understanding, seems the goal of value head is to estimate the KL divergence between policy and reference model so that they do not differ too much. And in PPO training, the entire reference network is not updated. My question is:

I am wondering, when initializing reference network with additional value head, is this value head randomly initialized? Also I think probably the value head is updated in training, otherwise the output won't make sense due to random initialization. If so, wondering what's the label used to train it? Very curious to learn some details behind it. 

I think without value head, we can use some likelihood to estimate pi_RL(y|x) and pi_SFT(y|x), but seems with it we can just use two scalars from value head to represent two probabilities. Wanted to learn if this understand is correct. Thank you so much!

**Comment by felipap** [comment]
The value head is trained during the PPO process itself. The loss used during PPO includes a term related to the value head, so during gradient descent (ascent?), the network also learns to estimate the correct values.

Best resource I found on PPO: https://www.youtube.com/watch?v=qGyFrqc34yc
